In [120]:
import pandas as pd
import os
import shutil

In [121]:
df = pd.read_csv("whole_data_fmri.csv")

In [122]:
df.head()

Image Data ID     Subject Group Sex  Age Visit Modality   
0      I1619408  941_S_7106   MCI   F   72    sc     fMRI  \
1      I1591326  941_S_7087    CN   M   67    sc     fMRI   
2      I1600185  941_S_7085   MCI   F   70    sc     fMRI   
3      I1588336  941_S_7074    CN   M   71    sc     fMRI   
4      I1553010  941_S_7051    CN   M   64    sc     fMRI   

                   Description      Type    Acq Date Format Downloaded  
0  Axial MB rsfMRI (Eyes Open)  Original  09-09-2022    DCM  5/29/2023  
1  Axial MB rsfMRI (Eyes Open)  Original   6/15/2022    DCM  5/29/2023  
2  Axial MB rsfMRI (Eyes Open)  Original  07-01-2022    DCM  5/30/2023  
3  Axial MB rsfMRI (Eyes Open)  Original  05-03-2022    DCM  5/30/2023  
4  Axial MB rsfMRI (Eyes Open)  Original  03-07-2022    DCM  5/30/2023

In [123]:
df.columns

Index(['Image Data ID', 'Subject', 'Group', 'Sex', 'Age', 'Visit', 'Modality',
       'Description', 'Type', 'Acq Date', 'Format', 'Downloaded'],
      dtype='object')

In [124]:
df['Group'].value_counts()

Group
CN         3005
EMCI       2205
LMCI       1346
MCI         904
AD          898
SMC         427
Patient       9
Name: count, dtype: int64

Group

CN 3005

EMCI 2205

LMCI 1346

MCI 904

AD 898

SMC 427

Patient 9


CN: 3006

EMCI: 2206
LMCI: 1347
MCI: 905
AD: 899
SMC: 428
PATIENT: 1


In [125]:
df['Group'].isnull().sum()

0

No Null Value in Group Column


Checking for any inconsistency in subject code


In [126]:
import re

# Define the regex pattern
pattern = re.compile(r'\d{3}_S_\d{4}')
test_string = "129_S_6858"
for subject in df['Subject']:
    if pattern.match(test_string):
        print("match")

match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
match
matc

NO there is no inconsistency in this column


Let's start segregation


In [127]:
grp_list = {'CN' : [], 
            'EMCI' : [],
            'LMCI' : [],
            'MCI' : [],
            'AD' : [],
            'SMC' : [],
            'Patient' : []}

In [128]:

root_dir = 'G:\ADNI\ADNI'
contents = os.listdir(root_dir)
new_dir = 'G:\ADNI\Segregated'
test_contents = contents
# print(len(test_contents))

In [134]:
for grp in grp_list:
    grp_list[grp].clear()

In [130]:
def traverse_folder(path):
    d = os.listdir(path)
    last_folder = None
    for item in d:
        item_path = os.path.join(path, item)
        if os.path.isdir(item_path):
            last_folder = traverse_folder(item_path)
        else:
            folders = os.path.normpath(path).split(os.path.sep)
            last_folder = folders[-1]
    return last_folder

In [140]:
for item in test_contents:
    subject_id = item
    item_path = os.path.join(root_dir,item)
    # print(item_path)
    fmri_id = traverse_folder(item_path)
    # print(fmri_id)
    for index, row in df.iterrows():
        subject = row['Subject']
        img_id = row['Image Data ID']
        group = row['Group']
        if subject == subject_id and img_id == fmri_id:
            if group in grp_list:
                    grp_list[group].append(subject_id)

In [142]:
for group, subjects in grp_list.items():
    print(f"{group}: {subjects}")


CN: [['002_S_0685'], ['002_S_1261'], ['006_S_6234'], ['006_S_6500'], ['014_S_4093'], ['023_S_0031'], ['023_S_1190'], ['024_S_4084'], ['024_S_4158'], ['029_S_4585'], ['029_S_6289'], ['032_S_6211'], ['032_S_6279'], ['036_S_4491'], ['037_S_4410'], ['067_S_0056'], ['067_S_0059'], ['070_S_4856'], ['070_S_5040'], ['128_S_4586'], ['128_S_4599'], ['130_S_4343'], ['131_S_6170'], ['131_S_6692'], ['135_S_6359'], ['135_S_6411'], ['941_S_6384'], ['941_S_6422'], ['941_S_6454'], ['941_S_6575'], ['941_S_6580'], '002_S_0685', '002_S_1261', '006_S_6234', '006_S_6500', '014_S_4093', '023_S_0031', '023_S_1190', '024_S_4084', '024_S_4158', '029_S_4585', '029_S_6289', '032_S_6211', '032_S_6279', '036_S_4491', '037_S_4410', '067_S_0056', '067_S_0059', '070_S_4856', '070_S_5040', '128_S_4586', '128_S_4599', '130_S_4343', '131_S_6170', '131_S_6692', '135_S_6359', '135_S_6411', '941_S_6384', '941_S_6422', '941_S_6454', '941_S_6575', '941_S_6580']
EMCI: [['009_S_2381'], ['011_S_4547'], ['024_S_2239'], ['029_S_23

In [144]:
for item in contents:
    item_path = os.path.join(root_dir,item) #sorce dir
    for catagory, subject_ids in grp_list.items():
        new_path = os.path.join(new_dir,catagory)
        if item in subject_ids:
            shutil.move(item_path,new_path)
